turbopack/crates/turbo-tasks-backend/src/backend/operation/mod.rs

```diff
@@ -445,11 +451,47 @@ impl<B: BackingStorage> TaskGuard for TaskGuardImpl<'_, B> {
     }
 
     fn add_new(&mut self, item: CachedDataItem) {
-        self.check_access(item.category());
-        let added = self.add(item);
+        let category = item.category();
+        self.check_access(category);
+        if !self.task_id.is_transient() && item.is_persistent() {
+            self.task.track_modification(category.into_specific());
+        }
+        let added = self.task.add(item);
         assert!(added, "Item already exists");
     }
 
+    fn extend(
+        &mut self,
+        ty: CachedDataItemType,
+        items: impl Iterator<Item = CachedDataItem>,
+    ) -> bool {
+        let category = ty.category();
+        self.check_access(category);
+        if !self.task_id.is_transient() && ty.is_persistent() {
+            let mut items = items.peekable();
+            // Check if the iterator is empty
+            if items.peek().is_none() {
+                return false;
+            }
+            // TODO this is not optimal as we always track a modification even if nothing is changed
```

Comment:
is it not sufficient to just use the return value of `extend` to decide if we need to track modifications?

Reply:
`track_modification` need to be called before changes are made as it might need to store the old value as snapshot.

Reply:
got it. thanks!